New Module: BCF switch

tedelhourani opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
bcf_switch

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (bcf 69f17bb39c) last updated 2017/03/13 19:08:53 (GMT -700)
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
Remove whitespace from unix group names

mattparkes committed:
ini_file 'create' parameter default value should be True

Updated create default value to True to match documentation (and common sense).

kbrebanov committed:
apk: Fix regex search patterns

Regex patterns were not being escaped properly so package names
containing characters that could be interpreted as regex symbols
were causing failures.

Fixes: #19714

kalgynirae committed:
Improve parsing of 'systemctl show' output

lpaulmp committed:
Open file in binary mode for python3

Commit:
Update the comments around Python 3

[Enterprise Linux 5 is at the end of "End of Production 3 phase"](https://en.wikipedia.org/wiki/Red_Hat_Enterprise_Linux#Product_life_cycle) so that is no longer a reasonable reason to wait.  Also removed the comments about [Python 3 not being mainstream](http://python3wos.mybluemix.net/201).  Perhaps a link to https://docs.ansible.com/ansible/python_3_support.html would also be a useful addition to this question.

kabergstrom committed:
Open unarchive file in binary mode

Recent Python3 versions require open() to specify binary mode if the data is anything other than text.

Python3: Use int() instead of long() in unarchive

Changes long() to int() for CRC values in the unarchive module. Affects unarchiving of zip files. Since CRC values in zipfile are 32 bits the behaviour should be unchanged even in Python 2.

jayakody commented:
ACK

jayakody commented:
LGTM. 
shipit

vuile commented:
ship

snowsky committed:
file module doesn't create hard link to given target directory

nerzhul committed:
New module: rundeck_project
This module permits to manage rundeck project creation or removal through rundeck API

nerzhul committed:
Add rundeck_acl_policy module

Commit:
stat: better documentation

  - md5 fails if the file is not a regular file ( symlink, dir,..)

bcoca committed:
fix hashing when path is symlink

bcoca committed:
fix environment validation, make setup exception

removed bare vars
now environment inheritance correclty prepends
this allows more local to override more general

fixes #23180

bcoca committed:
removed extra 'changed' from debug output

bcoca committed:
centralize ansible_managed and other template vars

now template lookup supports these again.

bcoca committed:
fix vars file selection

fixes #17382
alternate to #22979

deal with cases in which group/host have . in name
updated as per feedbck
only be strict about extension when doing dirs
also avoid ~ endings

abadger committed:
Fix a traceback with python3 and diff output

When retrieving file contents for diffing we need to get the contents as
binary.  Otherwise python3 will try to convert the file to text and fail
with non-decodable contents.

Fixes #23171

Qalthos committed:
Redo #23418 (#23425)

* Check for provider values inside check_args of respective network code

* Partial revert of b9ee5aa

The no_log change is okay, but take out the action result munging

umago committed:
rpm_key: Decode bytes to string to work with Python 3 (Fixes #20326) (#20385)

* rpm_key: Decode bytes to string to work with Python 3 (#20326)

The read() method will return bytes we need to then decode() those bytes
to a string before trying to match() it using the re module.

* Make the rpm_key pgp regex more robust on both py2 and py3

dagwieers committed:
vmware_guest_facts: Add delegate_to directive to example (#22931)

Without it, this is not going to do what most people expect.

This fixes #22840

Akasurde committed:
Fix atomic pylint issues

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

dmitry-sobolev committed:
Telegram notification module updated (#22858)

* msg_format parameter added
error message received from telegram API is added to fail json
compatibility with python3 added
* pep8 formatted
* version_added property added for msg_format
* bot token must be set without 'bot' prefix in module parameters
* formatting options described in documentation
* six module for compatibility used
telegram.py removed from legacy-files.txt

mattclay committed:
Fix ansible-test empty diff handling on Shippable.

mattclay committed:
Set IO encoding for ansible-test on Shippable.

indrajitr committed:
pacman: Fix package matching pattern in upgraded package list (#23433)

(cherry picked from commit 9a850db605a2db3c4f7f1cbc5bf88a8852ea17c4)

resmo committed:
add mailmap entries for @resmo

khnazaretyan committed:
Adds attached volumes to instance facts (#23132)

* Adds attached volumes to instance facts

* Fixed identation and modified coding style

* includes revisions after maintainers review

* Coding style modified

* simplifies getting items from dict

itdependsnetworks committed:
Update Banner Doc (#23412)

I think most people's banner will come from an external file. Show an example, and clarify there is no extra lines allowed.

LuckySB committed:
Fixes #23445

userlerueda committed:
Rebase using latest devel branch in order to fix conflicts (#23199)

Akasurde committed:
Pylint fixes for yum module

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

frenzymadness committed:
Add Python 3 support for validate-modules test. (#21195)

* Improve PEP8 compatibility

* Fix Python 3 incompatibility

Is prohibited to mutate OrderedDict during iteration through it so
is better to add records with error or warning to empty dictionary
instead of delete records from copy of dictionary during iterating.

* Decode output of subprocess from bytes to unicode.

* Add Python 3 support for validate-modules test.

Fix #18367

abadger committed:
When apt errors, include it's exit code

We invoke /usr/bin/apt inside of the ansible module.  When that command
exits, it doesn't always include a helpful error message.  Include the
exit code so that user's have all the information we can gice them as to
why apt failed.

Addresses #19128

bcoca committed:
Revert "made example not fail"

This reverts commit 2c9543e0b717d25fbe00b6085636941dda83d7a4.

example won't fail but will prevent correct PATH from being set
using actual fix to deal with this corner case

abadger committed:
Enable intersphinx for python stdlib and jinja2

This setting allows us to reference jinja2 and python stdlib docs
via docutils refs instead of hardcoding urls.

Fixes #18039

Remove extraneous sphinx config

abadger committed:
Add documentation of Jinja2 templating idioms that differ when run under Python3

Fixes #19514

acasademont committed:
Fix Postgres user module always reporting changes (#23488)

abadger committed:
Update version and change remove files from pep8 blacklist

gundalow committed:
Document how to debug network issues (#23503)

s-hertel committed:
[cloud] route53_zone: allow split horizon for route53_zone and refactor - fixes #22939 (#23190)

* allow split horizon for route53_zone and refactor

* fix documentation

remove comment

fix version_added

* Remove unused imports

* Only include zone as matching if it has the same privacy setting

* Use `.endswith` instead of indexing into a string

* Update public zone behavior to only create new if there is no matching public zone

* Remove from legacy PEP8 files

s-hertel committed:
[cloud] ec2_asg: check autoscaling group for tags before trying to use them - partially fixes #23234 (#23259)

* check autoscaling group for tags before trying to use them

* Add warning

fix comparison

abadger committed:
Ryanpineo fix ipaddr netmask with 32 cidr (#23506)

* Fix bug in ipaddr netmask filter when cidr is 32

* Add requirement on netaddr to the units.txt file

Commit:
Add checks to confirm that the output from `xfs_info` has the expected format.

jimi-c committed:
Adding 2.3.0 release date to release vars versions

abadger committed:
Add .coveragerc to the tarball so unittests will run from the tarball

jamielennox committed:
Allow modifying security groups in os_server (#23207)

When the security groups specified to the os_server module change they
should be updated on the server. This will require shade 1.19 where the
server security group commands were added.

Fixes: #23206

bcoca committed:
added test emulating #23180 use case

gundalow committed:
[WIP] Network return docs (#23441)

Network return docs

mattclay committed:
Include .yamllint in manifest so yamllint works.

brettcannon committed:
Fix reST markup (#23536)

pmarques committed:
fixes #4441 passing subnet_name parameter (#18920)

gundalow committed:
Use exports (#23542)

mattclay committed:
Fix module validator blacklist handling.

Process the blacklist before creating a validator instance.

mattclay committed:
Fix var precedence check to support python 3. (#23552)

* Fix var precedence check to support python 3.
* Run CI sanity tests using python 3.5.
* Disable pylint non-iterator-returned test to pass on python 3.5.

skg-net committed:
Error string:Regular expression updates (#23426)

bcoca committed:
less code

gundalow committed:
network_cli modules don't always return stdout or stdout_lines (#23573)

It's possible that if the module has a low-level failure, such as
"unable to open shell", or something else in the action plugin that
stdout and stdout_lines will not be returned.

Update the documentation to clarify this point.

mattclay committed:
Temporarily cap pylint version < 1.7.0 in tests.

gundalow committed:
Make it easier to find network modules (#23576)

* Make it easier to find network modules

Feedback has been it's difficult (via Google or directly) to find
modules as some people search for the company name vs product name,
therefore specify both.

* "IOS XR" (not "IOS-XR")

claer committed:
Improve Fortios IPv4 policy with logging capabilities and use the backup_filename param (#23544)

* Improve Fortios IPv4 policy with logging capabilities. While there, fix typos in examples. forti_config: use the backup_filename param and dont enforce the the filename value.

* forti-typos

* Add version_added for new options in the documentation

mattclay committed:
Add rstcheck to ansible-test and correct issues. (#23550)

* Add rstcheck to ansible-test.
* Fix rst code-block languages and syntax errors.
* Fix rst inline literals.
* Update python 2 code block to pass tests on py 3.

dagwieers committed:
Windows: Add Windows Subsystem for Linux documentation (#23515)

* Windows: Add Windows Subsystem for Linux documentation

As discussed during the Windows Working Group meeting we do want
documentation on running Ansible on Windows as it may help grow
the Windows/Ansible community and does work out-of-the-box.

However we do take care to emphasize that WSL is not fit for
production use.

* Update intro_windows.rst

Edits for mechanics and clarity.

luqasz committed:
Fix alarm action comparison (#23523)

This fixes issue when list from module contains more than one element.
Ansible and/or boto may put same elements in list in different order,
thus resulting task as changed.

Fixes #3310

dagwieers committed:
Windows: Get rid of Set-Attr in remaining modules (#23525)

Only a few more modules were using Set-Attr on the $result object rather
than using a normal hashtable. This PR changes the PSObject to a
hashtable and gets rid of Set-Attr.

Commit:
eu-west-1a is an availability zone, not a region. (#23557)

bcoca committed:
removed debug print

mattclay committed:
Expand test documentation. (#23598)

* Replace test README.md with new README.rst.
* Add compile and sanity README.rst docs.

privateip committed:
fixes issue where network modules would ignore provider host value (#23589)

This addresses a problem where the action plugin would ignore the
remote_addr value for the host.  In this case, only the inventory values
for the hostname would be considered and populate the remote host
remote_addr value for the connection plugin.

pilou- committed:
Debian: really generate locales

/usr/sbin/locale-gen differs between Debian and Ubuntu. With Debian,
locales must be added to /etc/locale.gen before calling locale-gen
command. With Ubuntu, /etc/locale.gen is updated by
/usr/sbin/locale-gen.

Use the locale-gen module which handles both distributions.

agaffney committed:
Allow specifying timeout on a per-host basis

manuel-sousa committed:
Fix #23338 - rabbitmq_queue message ttl can't be 0 (#23448)

Commit:
Fix documented defaults in the nsupdate module (#23532)

Let the documentation reflect what the module actually implements.

Akasurde committed:
Update package installation warning (#23641)

Github_release requires github3.py==1.0.0a4, update warning
for the same.

Fixes https://github.com/ansible/ansible/issues/23626

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

ganeshrn committed:
Make logging consistent for network platforms (#23605)

Qalthos committed:
Clean up import statements for nxos_rollback (#23540)

* Clean up import statements for nxos_rollback

* Don't try to deal with pre-2.3 logic

dgjustice committed:
Nxos vxlan vtep vni netcfg fix (#23620)

* Code cleanup
Removed 'add' method from CustomNetworkConfig.  It is identical
to the one inherited from NetworkConfig

* Removed unused CustomNetworkConfig import

* Replaced
```
def get_existing(module, args):
    existing = {}
    netcfg = get_config(module)
    config = netcfg.get_section(parents)
```
with
```
netcfg = CustomNetworkConfig(indent=2, contents=get_config(module))
```
get_config returns a string, not an object in 2.3.

* Removed non-functioning get_object method in CustomNetworkConfig in favor of the
inherited method.

Added child_objs property so that expand_selection would work.  The original
verion never worked correctly as it compared NetworkConfig obj's and str's.

* Removed ShellError method in favor or new load_config method.

* Removed ShellError method in favor or new load_config method.
fixes #20260

* nxos requires a "no" statement to change mcase group.  Corrected.
Corrected changed logic.

* Corrected deleted CustomNetworkConfig import

Qalthos committed:
pep8 fix for Shippable

itdependsnetworks committed:
Remove Warning for provider (#23652)

antoineco committed:
Fix nonexistent module_name in callbacks (#23587)

Spredzy committed:
crypto/openssl_csr: Specify proper Ansible version

The commit was started before 2.3 was branched, but was only merged once
2.3 was actually branched. This leads to documentation stating this
module is new in 2.3 when it will be actually new in 2.4

abadger committed:
display path to module_utils files in verbosity vvv.

dharmabumstead committed:
Switch Adobe tracking off staging.

dagwieers committed:
vmware_guest_snapshots: Add delegate_to directive to examples (#22933)

Without it it is not doing what most people would expect.

willthames committed:
[cloud] Add ec2_vpc_dhcp_options_facts check mode (#23106)

* Support check mode in ec2_vpc_dhcp_options_facts

As a facts module, ec2_vpc_dhcp_options_facts supports check mode
by default

* ec2_vpc_dhcp_options_facts tidy up

Use named method imports, move imports to top of code
Use shared code to handle filters and tags
Use snake case for parameter names while retaining backward compatibility

adejoux committed:
Added AIX LV module (#20574)

* Added AIX LVM module

* implemented requested changes

* fixed Terabytes size calculation

* implemented recommendations from tmshn

* moved up lv display command definition

* added type to copies parameter

* applied recommendations

* missed chlv command

* fixed build error

* fixed typo in email address

* applied new pep reco

* fix bug

dagwieers committed:
Add example task succeeding when RC is non-zero (#23698)

* Add example task succeeding when RC is non-zero

I added an example on how to use the return code to decide yourself
what is considered a failure.

This might have helped for #23679.

* Use diff as example command, instead of robocopy

Qalthos committed:
nxos_vlan tweaks (#23194)

* Remove commented code

Remove a lot of unnecessary output

* Tests

* Sort vlan ids _after_ running set operations, so order is not lost

bcoca committed:
added mongodb lookup docs

bcoca committed:
tolerate 'batch' systems that mess with stdin (#23596)

* tolerate 'batch' systems taht mess with stdin

fixes #23541

* have pause on windows tolerate devnull

* tuplie

shepdelacreme committed:
[cloud] Fix #23152 in route53 module and pep8 cleanup (#23156)

update module to support more standard state=present/absent syntax

update module to use required_if, required_together, mutually_exclusive functions where possible

per ryansb review: make documentation section more clear, fix some extra quotes, remove FIXME comment

pre willthames review: force private_zone to True if vpc_id is set and fix word wrap

ganeshrn committed:
Make warning logs consistent (#23666)

* Make warning logs consistent

Arguments outside provider with default
value should not log as warning in case
it is not mentioned in play.

* Make nxos timeout default consistent and add comments

* Make comments more verbose

sivel committed:
When become_method is su, self._play_context.prompt is a function. Fixes #23689

alikins committed:
Fix vault reading from stdin (avoid realpath() on non-links) (#23583)

* Fix vault reading from stdin (avoid realpath() on non-links)

os.path.realpath() is used to find the target of file paths that
are symlinks so vault operations happen directly on the target.

However, in addition to resolving symlinks, realpath() also returns
a full path. when reading from stdin, vault cli uses '-' as a special
file path so VaultEditor() will replace with stdin.
realpath() was expanding '-' with the CWD to something like
'/home/user/playbooks/-' causing errors like:

        ERROR! [Errno 2] No such file or directory: u'/home/user/ansible/-'

Fix is to specialcase '-' to not use realpath()

Fixes #23567

* to_text decrypt output when writing to stdout

bcoca committed:
nicer docs

s-hertel committed:
s3_sync: removing irrelevant s3.list_buckets() call - fixes #23409 (#23492)

* Removing irrelevant s3 call

Fix exception handling

Make s3_sync pep8 and remove from legacy file

dharmabumstead committed:
Update playbooks_python_version.rst

Minor edits

irlevesque committed:
Add ssh_connection / retries option documentation (#23463)

No docs were included when https://github.com/ansible/ansible/pull/9927 was merged.

gundalow committed:
Improve network debug wording (#23735)

stylersnico committed:
ConfigureRemotingForAnsible: RSA 1024 to RSA 4096  (#23684)

Akasurde committed:
Fix for setting values using gconftool-2 (#23713)

Miscallenous fixes added to make gconftool-2 work.

Fixes https://github.com/ansible/ansible/issues/23670

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

drew-russell committed:
Add the Cisco Spark Module to the 2.3 New Modules (#23636)

Update the Ansible 2.3 New Module list to include the cisco_spark notification module.

asys-gubi001 committed:
Fix example fact(ansible_default_ipv4.address) for delegated_facts

bcoca committed:
finer grained failed/skipped on async for actions

fixes #23729

abadger committed:
Really minor optimization and style change for d0e3d75381292fc7bf417598f9f6542a7fcd2d07

* Use a generator expression instead of a list comprehension
* Use copy() to create a new set instead of the constructor.

Commit:
fix ansible galaxy file mangling (#23703)

without this patch, ansible-galaxy will mangle files containing
the archive parent directory name, eg 'owncloud/files/owncloud.cron'
will become 'owncloud/files/.cron'.

The previous code could affect the entire path and even filenames.
If a file path has the top level dir name as a substring, galaxy
replaces it with ''. In one example, the archive top level dir
is 'go', so 'files/go-bin.sh' becomes 'files/-bin.sh'.

Fixes #22572, #23694, #23623

raszi committed:
Fix Markdown formatting for GitHub (#23752)

lattwood committed:
Use isfile check on package manager paths

Akasurde committed:
Minor typo in vmware_guest_facts (#23791)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

nerzhul committed:
Show UUID instead of name when using UUID to select snapshot (#23787)

trishnaguha committed:
Add eos_banner as module to banner param (#23789)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

rcarrillocruz committed:
Add missing aliases file to iosxr integration tests targets (#23796)

rcarrillocruz committed:
Add missing aliases file to ios_system integration target (#23795)

t0mk committed:
Add guide for Packet host (#21346)

* Add guide for Packet host

* incorporate feedback from Packet

* Format headings according to Python doc guide

* fixed rstcheck issues in packet guide

bcoca committed:
script fails on async

bcoca committed:
also skip polling when async failed

adamchainz committed:
Tidy example inventories in "Developing Dynamic Inventory Sources" (#18758)

Thanks @adamchainz!

ericsysmin committed:
changed code block formatting to match format (#23615)

babineaum committed:
Typo (#23778)

Missed that D...

abadger committed:
Facts Timeout was not settable via ansible.cfg

The timeout for gathering facts needs to be settable from three places
(highest precedence to lowest):

* programmatically
* ansible.cfg (equivalent to the user specifying it explicitly when
  calling setup)
* from the default value

The code was changed in b4bd6c80deccc64655e158cf4799edcf604c76d1 to
allow programmatically and the default value to work correctly but
setting via ansible.cfg/parameter was broken.

This change should fix setting via ansible.cfg and adds unittests for
all three cases

Fixes #23753

Qalthos committed:
EOS default transport (#23824)

* Give a default to eos transport

Qalthos committed:
Eos port default values (#23829)

* Hook port back up for eos cli transport

* The default values have moved

* Why not have conditional defaults?

s-hertel committed:
route53_facts: add example for how to use next_marker option - fixes #23625 (#23802)

* Add example for using route53 next_marker option

* make route53_facts pep8 and remove from legacy files

s-hertel committed:
improve exception handling (#23722)

make pep8

Akasurde committed:
Use regex variable in eos_user.py (#23797)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

bcoca committed:
less code

(cherry picked from commit 20d9f79a5250e764ffeddb046aeb82b7824ed952)

bcoca committed:
Revert "fixes play context connection user (#21776)" (#23830)

* Revert "fixes play context connection user (#21776)"

This reverts commit 58ee661437f123fdd8cf516bac053fe593910ced.
fixes #23530

* preserve original in copy

updated comment to clarify wtf is going on here

nitzmahone committed:
template hostvars for set_host_overrides (#23839)

* fixes #23586
* temporary solution until connection-specific vars are handled by play_context

fridim committed:
async+script: add integration test (#23818)

see #23729

ganeshrn committed:
Add doc for display argument of junos_command (#23804)

trishnaguha committed:
Add eos_user module to users parameter (#23847)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

rcarrillocruz committed:
Fix the iosxr_facts mem gathering (#23850)

We were not calling match.group, plus we were lacking a ':' from
the expected output of 'show memory summary'.

Fixes #23737

rcarrillocruz committed:
Add assertion on test_iosxr_facts (#23852)

We hit bug #23737 due to bad coverage on test_iosxr_facts, we
were not checking memory facts at all.

mscherer committed:
Fix --force for unversionned requirements (#23391)

In current stable (2.2), ansible galaxy install --force do erase
a role, even if the version is not set. This commit should restore
that specific behavior, in accordance to people reports:
  https://github.com/ansible/ansible/issues/11266#issuecomment-273801480

It was also the behavior planned in the initial discussion:
"if you're not fixing versions in your roles file, then it's fine
to expect that the role will be reinstalled each time you run
ansible-galaxy install.", cf https://github.com/ansible/ansible/pull/12904

ganeshrn committed:
Add junos_command module unit test (#23819)

* Add junos_command module unit test

ganeshrn committed:
Fix display argument defualt value (#23859)

rcarrillocruz committed:
Remove catch-all regex on IOS terminal plugin (#23858)

We have a list of specific messages that we scree-scrape and flag
them as legit errors.
However, we also have a catch-all regex that matches everything
starting with %.
That can cause issues on commands that return lines with that
character, like for example the 'crypto key generate'.

Fixes #23770

bcoca committed:
Add a new filter: strftime. Use the well known function to format a date output. (#23832)

(cherry picked from commit 3f5b304fc28b6c34df9ea6a4c3531dc422ce198b)
rebased for @yannig

bcoca committed:
keep unsafe .. unsafe (#23742)

* keep unsafe .. unsafe

fixes #23734, which was broken in previous fix that allowed non string types to be templated
use new 'is_template' function vs bastardizing others
refactored clean_data to allow for arbitrary data structures to clean
fixed/removed some tests

* deal with complex data for is_template

* typos

bcoca committed:
moved to exceptions for basic skip/fails

better handling of checkmode and async
fix test to follow new flow control

machacekondra committed:
cloud: ovirt: add note that modules works with RHV as well (#23808)

bcoca committed:
fix environment populated detection

fixes #23764

bcoca committed:
p8p

s-hertel committed:
make lineinfile docs clearer and make module pep8 (#23857)

dmahler committed:
Minor grammatical corrections and improvements (#23886)

jctanner commented:
bot_status

machacekondra committed:
cloud: ovirt: import VM from external source (#22450)

alikins committed:
Use sys.stdout.buffer to write vault bytes to stdout on py3 (#23760)

* Use sys.stdout.buffer to write vault bytes to stdout on py3

We need sys.stdout.buffer on py3 so we can write bytes to it since the plaintext
of the vaulted object could be anything/binary/etc

Before, attempting to write bytes to stdout on py3 would cause:

  TypeError: write() argument must be str, not bytes

kustodian committed:
Move include_vars to host scope (#23868)

* Move include_vars to host scope

Module include_vars actually sets variables on a host level, and not the play level.

* Expand vars entires

bcoca committed:
avoid hosts 'pre templated warnings' (#23878)

* avoid hosts 'pre templated warnings'

fixed is_template failing on undefined
fixes #23057

* py3 compat

s-hertel committed:
[cloud] ec2_facts: make ec2_facts python3 compatible - fixes #23595 (#23872)

* make ec2_facts python3 compatible

* remove parentheses and use pop() instead of del

s-hertel committed:
efs: include name in results (#23931)

indrajitr committed:
pacman: Fail fast when 'rc != 0'

We fail-fast and display 'stderr' in case 'pacman' returns with 'rc != 0'.
There is no point computing 'module._diff' in such case anyway.

Fixes #23910

(cherry picked from commit 8c6a2a848cf6a6d6522c8f5be56decf8df1ed6ab)

Commit:
Fix systemd in chroot (#23904)

* Fix systemd in chroot

The 'request ignored' message is in stderr, not stdout.

* Check both stdout and stderr for systemd message

Some versions of systemd report to stderr, others to stdout. Also check
whether output could be a valid normal response to avoid false positives.

bcoca committed:
fixed code to actually check the parameter not key

also made rest conditional on unit being provided, previouslly it changed from being always required

fixes #23810 alternate to #23907

trishnaguha committed:
Remove repetition of code nxos (#23908)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

trishnaguha committed:
Set terminal width for network *_command modules to avoid paging (#23728)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

resmo committed:
exoscale: fix TXT type records handling (#23956)

bcoca committed:
schedule sudo/su removal

dutchiechris committed:
Fix name description from lun to volume (#23964)

Help text stated to manage lun name but this module is for managing volumes.  Updated accordingly.

ganeshrn committed:
Fixes #23960 junos_config fail with config in xml (#23962)

If config is in xml format append it to <configuration>
tag as a instance of Element class.

ganeshrn committed:
Minor junos_config module changes (#23972)

* Minor junos_config module changes

* Add note for json config support
* Add example for lines argument
* Fix future warning for ElementTree

* Remove unwanted space

rcarrillocruz committed:
Unit tests for openvswitch_db module (#23974)

Brian-Williams committed:
Change wait_for time to utc (#23987)

erwilan committed:
synchronize: Fix (delegated) local rsync

Makes delegated local rsync work even if ansible_host or ansible_ssh_host is set.
Makes local rsync work when no ssh is installed.

keepzero committed:
Set validate_checksum default yes of fetch to match the doc

mattclay committed:
Fix module validator report if skipping all files.

ssidhaye committed:
Corrected -su cmdline option to --su to be consistent with the man page entry (#23899)

Signed-off-by: Sumedh Sidhaye <ssidhaye@redhat.com>

mattclay committed:
Use Shippable image: drydock/u16pytall:master (#23997)

* Use Shippable image: drydock/u16pytall:master
* Do not install python 3.6 on Shippable.

mattclay committed:
Update validate-modules to ignore *.rst files.

cptchips committed:
[Doc] win_package - product_id is a mandatory parameter (#24007)

product_id is a mandatory parameter for windows/win_package module.

supertom committed:
Added subnetwork parameter to inventory instance dictionary. (#23984)

Akasurde committed:
Pep8 fixes for mysql module (#23923)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
Minor typo and PEP8 fixes for dnf module (#23845)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

ka7 committed:
spelling fixes (non-trivial, might catch a bug) (#23646)

ogenstad committed:
asa_command: set default_output to text (#23979)

abadger committed:
fetch idempotence test and deprecate validate_md5

Added an integration test for fetch module idempotence.  (Testing
that validate_checksum is doing what it's supposed to is harder as we'd
have to create a race condition with the downloaded data to trigger it.
Probably need to make that a unittest eventually).

Also give a deprecation message to the validate_md5 parameter so that we
can eventually get rid of it.

Shaps committed:
fixed RETURN docs for modules (#24011)

* fixed RETURN docs for remaining modules

* updated proxymysql_mysql_users 'sample' to yaml dict

* fixed whitespace errors

giuseppe committed:
Atomic containers (#21366)

* atomic_image: add support for pull --storage=

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

* atomic_container: new module

Manage atomic containers

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

Akasurde committed:
Handle fetch_url error gracefully (#23821)

Fix adds http response code check for fetch_url.
Also, method in fetch_url call.

Fixes https://github.com/ansible/ansible/issues/23800

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Qalthos committed:
nxos_acl fixes (#23915)

* Update nxos_acl

* unit tests for nxos_acl

* Remove nxos_acl from pep8-legacy

Qalthos committed:
nxos_acl_interface fixes (#23917)

* Update nxos_acl_interface

* Add basic unit tests to nxos_acl_interface

Qalthos committed:
nxos_bgp fixes (#23927)

* Pre-emptively fix sanity/pep8

* fix for commands

* Quick unit test

jctanner committed:
Remove unused+circular import in listify.py (#24038)

abadger committed:
Fix circular import with unsafe_proxy, template, and vars

template/__init__.py imported unsafe_proxy from vars which caused
vars/__init__.py to load.  vars/__init__.py needed template/__init__.py
which caused issues.  Loading unsafe_proxy from another location fixes
that.

abadger committed:
Document deprecation of fetch module validate_md5 and update --tags merging deprecation (#24022)

* Document deprecation of fetch module validate_md5 and update --tags merging deprecation

Update the default of --tags merging config option to merge by default

* Update CHANGELOG.md

Minor edit

mattclay committed:
Fix bugs in ansible-test units command. (#24044)

* Handle old versions of coverage.
* Handle old versions of setuptools.
* Detect python version for docker/remote units.
* Add sanity override for test constraints.

anryko committed:
ecs_service: add tasks placement constraints and strategy. (#20503)

ecs_service: documentation fix.

ecs_service: updated version_added.

ecs_service: pep8 fixes.

ganeshrn committed:
Fix exception issue in junos_config (#24049)

ParseError execption is added to ElementTree
in py2.7. Prior to py2.7 need catch ExpatError
execption

kustodian committed:
Add a missing leading zero in a mode parameter

600 is an incorrect mode, because mode needs to be octal.

kustodian committed:
Remove action keyword from do-until example

`action:` is old and not needed any more.

skg-net committed:
Fixed PEP8 issues (#24030)

Shaps committed:
Fixed docs for cloud modules (#23745)

* Fixed docs for cloud modules

* removed blank line in ec2_snapshot_facts

rcarrillocruz committed:
Refactor openvswitch unit tests (#24055)

Rather than passing a file to load fixture, build a matrix containing
the run_command side_effect per test.
This will allow more code-reuse for other ovs modules unit tests.

Akasurde committed:
Pep8 fixes for known_hosts.py (#24029)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

rcarrillocruz committed:
Return commands instead of command (#24061)

Ok, so for openvswitch_db in particular we just return one command
but in the sake of consistency and code re-use, let's return a list
of "commands", even if it's just one.

ganeshrn committed:
Fix python version check issue (#24063)

Fix python version check issue

ganeshrn committed:
Add junos_config unit test (#24005)

Qalthos committed:
Run `save` inside config mode. (#23977)

* Run `save` before exiting config mode.

* Fix unit tests for `save`

* Allow `save` to be on its own again and introspect success

* Introspecting `compare running` makes this a lot harder.

Move `save` tests to integration tests

gundalow committed:
Split modules/network into two parts (#24024)

* Split modules/network into two parts

Given the dedicated team we we have working on Ansible Networking a
clearer split is needed between Networking modules and "things that
happen to use the network"

* nmcli to net_tools

* nmcli moved

Qalthos committed:
Junos_config unicode (#23369)

* Try to handle unicode output more sensibly

* Appears I'm getting latin1 instead

Ugh.

trishnaguha committed:
Refactor nxos_interface and add unit test (#24008)

* [WIP] Refactor nxos_interface

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

* Modification in refactor
Unittest for nxos_interface

gundalow committed:
Added note on netcat dependency (#24070)

Originally https://github.com/ansible/ansible/pull/23911/files

gundalow committed:
nmcli: fix docs for add necessary package (#24071)

Originally report in https://github.com/ansible/ansible/pull/23817

bcoca committed:
added versions to dep notices (#24019)

* added versions to dep notices

* pep7

* string

gundalow commented on lib/ansible/modules/network/bigswitch/bcf_switch.py:
```diff
@@ -0,0 +1,183 @@
+short_description: Create and remove a bcf switch.
+description:
+    - Create and remove a bcf switch.
+version_added: "2.3"
```
`version_added: "2.4"`

gundalow commented:
Closing and reopening to rerun CI

mattclay committed:
Fix PEP 8 issue.

pilou- committed:
PostgreSQL Integration test suite: improvements (#23631)

* PostgreSQL tests: use package module
* PostgreSQL tests: use ansible_service_mgr in order to test init system
* PostgreSQL tests: fix a typo
* PostgreSQL tests: install only required packages
* PostgreSQL tests: use role default value when unchanged

trishnaguha committed:
Fixes #23890 Remove idle ShellError exception and get_exception() method from nxos modules (#23947)

Remove idle ShellError exception and get_exception() method from nxos modules

gundalow committed:
Docs how to test (2nd) (#24094)

* Big testing doc refactor
* Combine all the testing documentation in to one place to make it easier to find
* Convert everything to RST
* Create testing_network guide
* Create testing landing page
* For each section detail "how to run" and "how to extend testing"
* More examples
* Lots more detail

pipern committed:
Minor typo correction varibles -> variables

No impact as variable wasn't used.

gundalow committed:
Minor testing docs improvements (#24103)

Fix a few formatting issues spotted post review.
Also reapply missing commit

gundalow committed:
Correct RETURN data types for Network modules (#24104)

`path` isn't a datatype

Akasurde committed:
Pep8 fixes for ha_proxy module (#24090)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

SesquipedalianDefenestrator committed:
Add check_mode to get_url  (#20532)

* Add check_mode to get_url that does a HEAD request to make sure the URL exists, but doesn't write the real file

* Add info about new --check behavior to docs.  Add tests for the new behavior.  Populate res_args with the info the tests are looking for.

* Add trailing comma

* Change nonexistent test URL to http://{{httpbin_host}/DOESNOTEXIST.  Fix spacing while I'm at it

* Further spacing cleanup

* State that this functionality is in Ansible 2.4+

Commit:
Rename fact returned by docker_network to avoid restricted prefix (#23919)

This fixes #23918

ganeshrn committed:
Fix exception in junos_package (#24107)

rcarrillocruz committed:
Refactor openvswitch_bridge (#24014)

Use common patterns from other network modules for better
maintainability.

gundalow committed:
dict, not dictionary (#24123)

Qalthos committed:
Print specific error for BadHostKeyException (#24120)

abadger committed:
Remove hack for backwards compatible v2_playbook_on_start callback

Just after release of 2.0.0 (in 2.0.0.1) we had a change to the API of
callbacks without bumping the API version.  We added the playbook to the
arguments passed to the callbacks.

This wasn't in the Tower callback at the time.  In order to prevent
breaking that callback we added a temporary hack to inspect the
callback's API to decide if we needed to call it with arguments or not.

We scheduled the hack for removal in January 2017.  Since that's now
past, removing the hack.

Change signed off by matburt on the Tower side.

cboggs committed:
Fix orphan resource pool crash (Issue #24078) (#24079)

Fixes issue [#24078](https://github.com/ansible/ansible/issues/24078)

Resource pools are retrieved from VSphere regardless of the state of the `resource_pool` parameter. During this process, each retrieved Resource Pool is checked for the `parent` attribute, and if it exists, the currently-scoped parent object is compared against the Resource Pool's parent object. The method doing the check, however, `assert`s that the parent object is not `None`. In some cases, a Resource Pool will have the `parent` attr, but that `parent` object will be `None`, causing the `assert` to fail.

This should avoid that. :-)

Commit:
Add support for specifying the modulepath for the Puppet module (#24122)

Puppet modules are not always installed in the default location
(i.e, /etc/puppet/modules) so it is useful to be able to specify
an alternate location.

pchauncey committed:
Add 'uuid' to the attributes returned from the vmware_vm_facts module. (#23773)

dankolbrs committed:
Fixes minor type in docs (#24117)

Removes an extra `e` from the word `batch` in the the
`playbooks_delegation` documentation seen at:
http://docs.ansible.com/ansible/playbooks_delegation.html

Commit:
Fix ClearLinux pretty name

ClearLinux has changed its pretty name in os-release file
from: 'Clear Linux Software for Intel Architecture'
to: 'Clear Linux OS for Intel Architecture'

This patch makes the SEARCH_STRING 'Clear Linux' rather than
the full name to make it compatible with the old and new name.

Signed-off-by: Alberto Murillo Silva <alberto.murillo.silva@intel.com>

resmo committed:
cloudstack: cs_region: fix pep8 (#24134)

resmo committed:
rundeck_project: set ansible_metadata status to preview

Qalthos committed:
Remove `provider` requirement from nxos_vrf (#24118)

Fixes #23920

sivel committed:
Speed up env-setup (#24133)

Akasurde committed:
Update documentation of fetch_url in urls.py (#23892)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
Pep8 fixes for taiga_issue module (#24143)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
Pep8 fixes for jenkins_* module (#24154)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
Misc fixes and Pep8 fixes for proxmox_* modules (#24162)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

sivel committed:
Fix password prompt matching (#24081)

* Fix password prompt matching

* Add some tests for check_password_prompt

* Prevent pep8 line ends with a space error

Akasurde committed:
Pylint fixes for packaging (#23553)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

abadger committed:
Add integration test that modules are loaded from the expected locations (#24170)

* Add integration test that modules are loaded from the expected locations
* Fix ping module to pass PEP8

bcoca committed:
added epub entry for makefile

thomasbilk committed:
Added Composer Command global (#24158)

* Added Composer Command global

Added a parameter to run composer commands globally. The `working_dir`
parameter is only required if `global_command` is `False`.

Fixes #24052

* Added Composer Command global

Added a parameter to run composer commands globally. The `working_dir`
parameter is only required if `global_command` is `False`.

Fixes #24052

gundalow commented on lib/ansible/modules/network/bigswitch/bigmon_chain.py:
```diff
@@ -2,7 +2,7 @@
 # -*- coding: utf-8 -*-
```
Modules should have permission 644 (not 775), could you please check this.

gundalow commented on lib/ansible/modules/network/bigswitch/bcf_switch.py:
```diff
@@ -0,0 +1,185 @@
+---
+module: bcf_switch
+author: "Ted (@tedelhourani)"
+short_description: Create and remove a bcf switch.
```
One bit of feed back we've recently had fun the community is to put the vendor and/or product name in the `short_description` so a more searchable string will appear in http://docs.ansible.com/ansible/list_of_network_modules.html

So I think bigmon needs adding to this.

tedelhourani replied:
This is a bcf (big cloud fabric), not a big mon module.

gundalow commented on lib/ansible/modules/network/bigswitch/bcf_switch.py:
```diff
@@ -0,0 +1,185 @@
+            validate_certs=dict(type='bool', default='True'),
+            access_token=dict(type='str', no_log=True)
+        )
+    )
```
What would it take for the module to support "check mode", ie be idempotent?

tedelhourani replied:
Can be implemented now as the module is already idempotent.

But before, what does this mean?

lib/ansible/modules/network/bigswitch/bcf_switch.py:0:0: E319 RETURN.changed.: expected a dictionary
lib/ansible/modules/network/bigswitch/bcf_switch.py:0:0: E319 RETURN.invocation.description: required key not provided @ data['description']. Got None
lib/ansible/modules/network/bigswitch/bcf_switch.py:0:0: E319 RETURN.invocation.module_args: extra keys not allowed @ data['module_args']. Got {'validate_certs': False, 'fabric_role': 'leaf', 'controller': '192.168.86.230', 'mac': '00:00:00:02:00:02', 'name': 'Rack1Leaf1', 'state': 'present', 'access_token': None, 'leaf_group': 'R1'}
lib/ansible/modules/network/bigswitch/bcf_switch.py:0:0: E319 RETURN.invocation.returned: required key not provided @ data['returned']. Got None
lib/ansible/modules/network/bigswitch/bcf_switch.py:0:0: E319 RETURN.invocation.type: required key not provided @ data['type']. Got None
click here for bot help

gundalow committed:
Return type is dict, not dictionary (#24191)

Shaps committed:
Added test for 'RETURN' field in validate-modules (#23322)

* Added test for 'RETURN' field in validate-modules

* print the field being tested.

Useful when the RETURN structure is complex.

* Fixed schema after CI traceback fail

* Fixed list_string_types

* Fixed line in 319 code for RETURN

Akasurde committed:
Pep8 fixes for apache2 module (#24185)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

s-hertel committed:
sns: move from notification directory to cloud/amazon directory - fixes #11874 (#23805)

* move sns to cloud directory

* make sns pep8 and remove from legacy files

s-hertel committed:
elasticache_parameter_group: fix documentation and exception handling - fixes #23709 (#23718)

* fix documentation and correct exception handling

* follow AWS exception guidelines

* fix parameter_group_family req; only needed when creating cache parameter group

make pep8 and remove from legacy files

rcarrillocruz committed:
Sanitize ovs br-get-external-id output on openvswitch_bridge (#24209)

If a bridge does not have external_ids, ovs-vsctl returns '{}'.
This causes issues on the current want vs have comparison in cases
where the play does not define external_ids, as the comparison
is None vs '{}'.

rcarrillocruz committed:
Sanitize the output from ovs get-fail-mode (#24208)

If a bridge does not have a fail mode set, it returns nothing, i.e.
empty string.
This causes a failure when doing the want vs have compare in plays
where the fail-mode is missing, as we compare "" vs None respectively.

higginsd committed:
fixed missing "import traceback" (#23427)

robinro committed:
Make git module pep8 compliant (#24196)

Akasurde committed:
Pep8 fixes for django_manage (#24198)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
Pep8 fixes for letsencrypt module (#24144)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

mattclay committed:
Allow empty string for EC2 session token from env.

mattclay committed:
Fix ec2_elb_lb error handling. (#24183)

- Use `fail_json` for `get_all_load_balancers` unhandled exception.
- Do not mask error message on failure of `enable_zones`.
- Do not mask error message on failure of `disable_zones`.

Also added/removed blank lines to improve PEP 8 compliance.

tedelhourani committed:
Merge branch 'devel' into bcf

tedelhourani commented:
Closing this one starting from scratch.
